timewarp_gl/plugin.cpp

```diff
@@ -461,59 +426,31 @@ class timewarp_gl : public threadloop {
 		// TODO: Right now, this samples the latest pose published to the "pose" topic.
 		// However, this should really be polling the high-frequency pose prediction topic,
 		// given a specified timestamp!
 		const fast_pose_type latest_pose = pp->get_fast_pose();
-		GetViewMatrixFromPose(&viewMatrixBegin, latest_pose.pose);
-
-		// std::cout << "Timewarp: old " << most_recent_frame->render_pose.pose << ", new " << latest_pose->pose << std::endl;
+		viewMatrixBegin.block(0,0,3,3) = latest_pose.pose.orientation.toRotationMatrix();
 
 		// TODO: We set the "end" pose to the same as the beginning pose, because panel refresh is so tiny
 		// and we don't need to visualize this right now (we also don't have prediction setup yet!)
 		viewMatrixEnd = viewMatrixBegin;
 
-		// Get HMD view matrices, one for the beginning of the
-		// panel refresh, one for the end. (This is set to a 0.1s panel
-		// refresh duration, for exaggerated effect)
-		//GetHmdViewMatrixForTime(&viewMatrixBegin, glfwGetTime());
-		//GetHmdViewMatrixForTime(&viewMatrixEnd, glfwGetTime() + 0.1f);
-
-		//ksAlgebra::ksMatrix4x4f_CreateRotation( &viewMatrixBegin, (cursor_y - SCREEN_HEIGHT/2) * 0.05, (cursor_x - SCREEN_WIDTH/2) * 0.05, 0.0f );
-		//ksAlgebra::ksMatrix4x4f_CreateRotation( &viewMatrixEnd, (cursor_y - SCREEN_HEIGHT/2) * 0.05, (cursor_x - SCREEN_WIDTH/2) * 0.05, 0.0f );
-
 		// Calculate the timewarp transformation matrices.
 		// These are a product of the last-known-good view matrix
 		// and the predictive transforms.
-		ksAlgebra::ksMatrix4x4f timeWarpStartTransform4x4;
-		ksAlgebra::ksMatrix4x4f timeWarpEndTransform4x4;
-
-		// DEMONSTRATION:
-		// Every second, toggle timewarp on and off
-		// to show the effect of the reprojection.
-		/*
-		if(glfwGetTime() < 9.0){
-			viewMatrixBegin = viewMatrix;
-			viewMatrixEnd = viewMatrix;
-		}
-		*/
+		Eigen::Matrix4f timeWarpStartTransform4x4;
+		Eigen::Matrix4f timeWarpEndTransform4x4;
 
 		// Calculate timewarp transforms using predictive view transforms
 		CalculateTimeWarpTransform(&timeWarpStartTransform4x4, &basicProjection, &viewMatrix, &viewMatrixBegin);
 		CalculateTimeWarpTransform(&timeWarpEndTransform4x4, &basicProjection, &viewMatrix, &viewMatrixEnd);
 
-		// We transform from 4x4 to 3x4 as we operate on vec3's in NDC space
-		ksAlgebra::ksMatrix3x4f timeWarpStartTransform3x4;
-		ksAlgebra::ksMatrix3x4f timeWarpEndTransform3x4;
-		ksAlgebra::ksMatrix3x4f_CreateFromMatrix4x4f( &timeWarpStartTransform3x4, &timeWarpStartTransform4x4 );
-		ksAlgebra::ksMatrix3x4f_CreateFromMatrix4x4f( &timeWarpEndTransform3x4, &timeWarpEndTransform4x4 );
-
-		// Push timewarp transform matrices to timewarp shader
-		glUniformMatrix3x4fv(tw_start_transform_unif, 1, GL_FALSE, (GLfloat*)&(timeWarpStartTransform3x4.m[0][0]));
-		glUniformMatrix3x4fv(tw_end_transform_unif, 1, GL_FALSE,  (GLfloat*)&(timeWarpEndTransform3x4.m[0][0]));
+		glUniformMatrix4fv(tw_start_transform_unif, 1, GL_FALSE, (GLfloat*)(timeWarpStartTransform4x4.data()));
```

Comment:
I never really understood why 3x4 matrices were used in the first place, and now I don't understand why we can work with 4x4s. Can you clarify?

Reply:
This is a good one. I wasn't really sure either, but then I went down a long rabbit hole of linear algebra.

Previously, Van Waveren's code was using this weird GLSL hack where you could perform a transpose operation by multiplying the matrix on the right instead of the left. GLSL "flexibly" interprets vectors as either rows or columns depending on which side you multiply the matrix on; this allows you to do an implicit transpose operation on the matrix. 3x4 matrices were used because it was a right-multiplication on a vec3.

This change is essentially just removing this weird hack and doing all of the algebra "properly". It may be removing some tiny performance optimization (because 3x4 uses one fewer vector register, or something like that), but as far as I can read online, using 4x4 instead of 3x4 (i.e. removing the micro-optimization) has little effect, if any.

In my opinion, using standard 4x4 matrices with a standard multiplication order instead of the weird hack that was implemented previously has benefits for readability, maintainability, and extensibility. Especially if we are publishing this to be hacked upon and improved by others!

Reply:
Thanks for the explanation!